File: 3776.minimum-moves-to-balance-circular-array.java
#
# @lc app=leetcode id=3776 lang=java
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
class Solution {
    public long minMoves(int[] balance) {
        int n = balance.length;
        long total = 0;
        for (int b : balance) total += b;
        if (total < 0) return -1;
        int negIdx = -1;
        for (int i = 0; i < n; ++i) {
            if (balance[i] < 0) {
                negIdx = i;
                break;
            }
        }
        if (negIdx == -1) return 0;
        // Try both clockwise and counterclockwise
        long minMoves = Long.MAX_VALUE;
        // Clockwise
        long[] arr = new long[n];
        for (int i = 0; i < n; ++i) arr[i] = balance[i];
        long moves = 0;
        long needed = -arr[negIdx];
        int idx = (negIdx + 1) % n;
        while (needed > 0 && idx != negIdx) {
            long give = Math.min(arr[idx], needed);
            moves += give * ((idx - negIdx + n) % n);
            arr[idx] -= give;
            needed -= give;
            idx = (idx + 1) % n;
        }
        if (needed == 0) minMoves = moves;
        // Counterclockwise
        arr = new long[n];
        for (int i = 0; i < n; ++i) arr[i] = balance[i];
        moves = 0;
        needed = -arr[negIdx];
        idx = (negIdx - 1 + n) % n;
        while (needed > 0 && idx != negIdx) {
            long give = Math.min(arr[idx], needed);
            moves += give * ((negIdx - idx + n) % n);
            arr[idx] -= give;
            needed -= give;
            idx = (idx - 1 + n) % n;
        }
        if (needed == 0) minMoves = Math.min(minMoves, moves);
        return minMoves == Long.MAX_VALUE ? -1 : minMoves;
    }
}
# @lc code=end

File: 3776.minimum-moves-to-balance-circular-array.go
#
# @lc app=leetcode id=3776 lang=golang
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
func minMoves(balance []int) int64 {
    n := len(balance)
    total := int64(0)
    for _, v := range balance {
        total += int64(v)
    }
    if total < 0 {
        return -1
    }
    negIdx := -1
    for i, v := range balance {
        if v < 0 {
            negIdx = i
            break
        }
    }
    if negIdx == -1 {
        return 0
    }
    // Rotate so that negative index is at start
    rotated := make([]int, n)
    for i := 0; i < n; i++ {
        rotated[i] = balance[(negIdx+i)%n]
    }
    prefix := int64(0)
    moves := int64(0)
    for i := 0; i < n-1; i++ { // last one doesn't need to check
        prefix += int64(rotated[i])
        moves += absInt64(prefix)
    }
    return moves
}

func absInt64(x int64) int64 {
    if x < 0 {
        return -x
    }
    return x
}
# @lc code=end